extensions/amp-cld-img/0.1/utils.js

```diff
@@ -0,0 +1,376 @@
+      resourceType = 'authenticated_images';
+      type = null;
+    } else {
+      throw new Error('URL Suffix only supported for image/upload, ' +
+        'image/private, image/authenticated');
+    }
+  }
+  if (useRootPath) {
+    if ((resourceType === 'image' && type === 'upload') ||
+      (resourceType === 'images' && (type == null))) {
+      resourceType = null;
+      type = null;
+    } else {
+      throw new Error('Root path only supported for image/upload');
+    }
+  }
+  if (shorten && resourceType === 'image' && type === 'upload') {
+    resourceType = 'iu';
+    type = null;
+  }
+  return [resourceType, type];
+}
+
+/**
+ *
+ * @param {?string} source
+ * @param {?string} format
+ * @param {?string} urlSuffix
+ * @return {Array<string>}
+ */
+function finalizeSource(source, format, urlSuffix) {
+  let sourceToSign;
+  source = source.replace(/([^:])\/\//g, '$1/');
+  if (source.match(/^https?:\//i)) {
+    source = smartEscape(source);
+    sourceToSign = source;
+  } else {
+    source = encodeURIComponent(decodeURIComponent(source)).replace(/%3A/g, ':')
+        .replace(/%2F/g, '/');
+    sourceToSign = source;
+    if (!!urlSuffix) {
+      if (urlSuffix.match(/[\.\/]/)) {
+        throw new Error('urlSuffix should not include . or /');
+      }
+      source = source + '/' + urlSuffix;
+    }
+    if (format != null) {
+      source = source + '.' + format;
+      sourceToSign = sourceToSign + '.' + format;
+    }
+  }
+  return [source, sourceToSign];
+}
+
+/**
+ *
+ * @param {string} string
+ * @param {RegExp} unsafe
+ * @return {string}
+ */
+function smartEscape(string, unsafe = /([^a-zA-Z0-9_.\-\/:]+)/g) {
+  return string.replace(unsafe, function(match) {
+    return match.split('').map(function(c) {
+      return '%' + c.charCodeAt(0).toString(16).toUpperCase();
+    }).join('');
+  });
+}
+
+
+/**
+ * Generate a transformation string based on the options
+ * @param {Object} options
+ * @return {string}
+ */
+function generateTransformationString(options) {
+  const width = optionConsume(options, 'transformationWidth',
+      optionConsume(options, 'width', null));
+  const height = optionConsume(options, 'transformationHeight',
+      optionConsume(options, 'height', null));
+  const crop = optionConsume(options, 'crop', null);
+  const gravity = optionConsume(options, 'gravity', null);
+
+  let background = optionConsume(options, 'background', null);
+  background = background && background.replace(/^#/, 'rgb:');
+  const effect = optionConsume(options, 'effect', null);
+  const border = optionConsume(options, 'border', null);
+  const aspectRatio = optionConsume(options, 'aspectRatio', null);
+  const dprValue = optionConsume(options, 'dpr', null);
+  const rawTransformation = optionConsume(options, 'rawTransformation', null);
+  const fetchFormat = optionConsume(options, 'fetchFormat', null);
+  const quality = optionConsume(options, 'quality', null);
+
+  const transformation = {
```

Comment:
I agree it's a bit confusing, however the separation is there because each segment of the is processed separately - [raw_transformation/transformations fields/responsive fields].
It ends up looking like so (assuming no raw transformation, for simplicity):
`b_red,e_sepia/w_100,c_crop`
If all the fields are in the same object I can't really separate which fields come before the slash and which ones after, it can change the meaning completely.